### PR TITLE
fix: CometExecRule should handle ShuffleQueryStage and ReusedExchange

### LINKED ISSUE #185: ShuffleQueryStage and ReusedExchange are not processed by CometExecRule
### Describe the bug

I found that Sort in some TPCDS queries are not transformed to CometSort, e.g.,

```
+- * SortMergeJoin Inner (19)
   :- * Sort (14)
   :  +- * ColumnarToRow (13)
   :     +- ShuffleQueryStage (12), Statistics(sizeInBytes=16.1 GiB, rowCount=7.20E+8)
   :        +- CometColumnarExchange (11)
   :          +- CometFilter (10)
   :            +- CometBatchScan iceberg.1000g_parquet_zstd_512mb_block_partitioned_iceberg.web_sales (9)
   : - * Sort (18)
      +- * ColumnarToRow (17)
        +- ShuffleQueryStage (16), Statistics(sizeInBytes=16.1 GiB, rowCount=7.20E+8)
          +- ReusedExchange (15)
```

It is blocked by `ShuffleQueryStage` and/or `ReusedExchange` which `CometExecRule` doesn't handle currently.

We should handle these operators in `CometExecRule`.

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #185.

### CODE REVIEW

File: spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala [hunk 1]
```diff
@@ -64,8 +64,9 @@ class CometNativeShuffleSuite extends CometTestBase with AdaptiveSparkPlanHelper
           val path = new Path(dir.toURI.toString, "test.parquet")
           makeParquetFileAllTypes(path, dictionaryEnabled = dictionaryEnabled, 1000)
           var allTypes: Seq[Int] = (1 to 20)
-          if (isSpark34Plus) {
-            allTypes = allTypes.filterNot(Set(14, 17).contains)
+          if (!isSpark34Plus) {
+            // TODO: Remove this once after https://github.com/apache/arrow/issues/40038 is fixed
+            allTypes = allTypes.filterNot(Set(14).contains)
           }
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala:70]: Previously these columns fail on Spark 3.4. I just ran test with them and looks okay now.

Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala:70]: Besides, after this change, Comet native operator can be after `CometExchange`. So it triggers the known bug in Java Arrow on column `_14`. I exclude the column for Spark 3.2 and 3.3. 

```
== Physical Plan ==                                                                                                                                                                                                                    
AdaptiveSparkPlan isFinalPlan=true                     
+- == Final Plan ==                                                                                                
   *(1) ColumnarToRow                                                                                              
   +- CometProject [_1#556], [_1#556]                                                                              
      +- ShuffleQueryStage 0                                                                                                                                                                                                           
         +- CometExchange hashpartitioning(_13#568, 10), REPARTITION_BY_NUM, CometNativeShuffle, [plan_id=838]                                                                                                                         
            +- CometScan parquet [_1#556,_13#568] Batched: true, DataFilters: [], Format: CometParquet, Location: InMemoryFileIndex(1 paths)[file:/Users/liangchi/repos/arrow-datafusion-comet/spark/target/tmp/spa..., PartitionFilter
s: [], PushedFilters: [], ReadSchema: struct<_1:boolean,_13:string>                                       
+- == Initial Plan ==                                                                                              
   CometProject [_1#556], [_1#556]                                                                                 
   +- CometExchange hashpartitioning(_13#568, 10), REPARTITION_BY_NUM, CometNativeShuffle, [plan_id=830]
      +- CometScan parquet [_1#556,_13#568] Batched: true, DataFilters: [], Format: CometParquet, Location: InMemoryFileIndex(1 paths)[file:/Users/liangchi/repos/arrow-datafusion-comet/spark/target/tmp/spa..., PartitionFilters: [],
 PushedFilters: [], ReadSchema: struct<_1:boolean,_13:string>
```

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 2]
```diff
@@ -378,6 +382,27 @@ class CometSparkSessionExtensions
             case None => b
           }
 
+        case s @ ShuffleQueryStageExec(_, _: CometShuffleExchangeExec, _) =>
```
Comment (snmvaughan) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:385]: I'd suggest highlighting the difference between these 2 blocks with comments.  The actions are exactly the same, so the only difference is the case check.

**Review by snmvaughan — CHANGES_REQUESTED** [review]

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala (hunk 2 above)
Comment (viirya) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:385]: Okay

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala [hunk 3]
```diff
@@ -933,6 +933,54 @@ class CometShuffleSuite extends CometColumnarShuffleSuite {
...
+
+  test("Comet native operator after ShuffleQueryStage") {
+    withSQLConf(
+      SQLConf.ADAPTIVE_AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
+      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
+      CometConf.COMET_EXEC_ENABLED.key -> "true",
+      CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true",
+      CometConf.COMET_COLUMNAR_SHUFFLE_ENABLED.key -> "true") {
+      withParquetTable((0 until 10).map(i => (i, i % 5)), "tbl_a") {
+        withParquetTable((0 until 10).map(i => (i % 10, i + 2)), "tbl_b") {
```
Comment (parthchandra) [inline spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:947]: I think this is not needed in this test

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 4]
```diff
@@ -222,12 +222,16 @@ class CometSparkSessionExtensions
...
     // spotless:on
     private def transform(plan: SparkPlan): SparkPlan = {
-      def transform1(op: UnaryExecNode): Option[Operator] = {
-        op.child match {
-          case childNativeOp: CometNativeExec =>
-            QueryPlanSerde.operator2Proto(op, childNativeOp.nativeOp)
-          case _ =>
-            None
+      def transform1(op: SparkPlan): Option[Operator] = {
+        val allNativeExec = op.children.map {
```
Comment (parthchandra) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:226]: I don't see why this change is needed here.

Comment (viirya) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:226]: They are not UnaryExecNode.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala (hunk 3 above)
Comment (viirya) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:947]: Yea, I copied it from other PR and modified it. I will remove it.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 5]
```diff
@@ -378,6 +382,31 @@ class CometSparkSessionExtensions
...
+            case Some(nativeOp) =>
+              CometSinkPlaceHolder(nativeOp, s, s)
+            case None =>
+              s
+          }
+
+        // For AQE shuffle stage on a reused Comet shuffle exchange
+        // Note that we don't need to handle `ReusedExchangeExec` for non-AQE case, because
+        // the query plan won't be re-optimized/planned in non-AQE mode.
+        case s @ ShuffleQueryStageExec(
```
Comment (sunchao) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:398]: curious if we can replace this rule with:
```scala
        case ReusedExchangeExec(_, op) => op
```
?

**Review by sunchao — COMMENTED** [review]
Looks good

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 6]
```diff
@@ -222,12 +222,16 @@ class CometSparkSessionExtensions
      */
     // spotless:on
     private def transform(plan: SparkPlan): SparkPlan = {
-      def transform1(op: UnaryExecNode): Option[Operator] = {
-        op.child match {
-          case childNativeOp: CometNativeExec =>
-            QueryPlanSerde.operator2Proto(op, childNativeOp.nativeOp)
-          case _ =>
-            None
+      def transform1(op: SparkPlan): Option[Operator] = {
```
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:225]: These operators are not UnaryExecNode.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala (hunk 5 above)
Comment (viirya) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:398]: Hmm, then users won't see `ReusedExchangeExec` anymore in explain string or Spark UI. It is useful to know which part is reused by Spark.

Comment (sunchao) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:398]: oops you are right

**Review by sunchao — APPROVED** [review]

**Comment by viirya** [comment]
Merged. Thanks.

**Pull request merged**